File: kirun-java/src/main/java/com/fincity/nocode/kirun/engine/util/json/ConvertorUtil.java
package com.fincity.nocode.kirun.engine.util.json;

import static com.fincity.nocode.kirun.engine.json.schema.validator.reactive.ReactiveSchemaValidator.path;

import java.util.List;

import com.fincity.nocode.kirun.engine.json.schema.Schema;
import com.fincity.nocode.kirun.engine.json.schema.convertor.enums.ConversionMode;
import com.fincity.nocode.kirun.engine.json.schema.convertor.exception.SchemaConversionException;
import com.google.gson.JsonElement;
import com.google.gson.JsonNull;

public class ConvertorUtil {

	public static JsonElement handleUnConvertibleValue(List<Schema> parents, ConversionMode mode, JsonElement element,
			String errorMessage) {
		return handleUnConvertibleValue(parents, mode, element, JsonNull.INSTANCE, errorMessage);
	}

	public static JsonElement handleUnConvertibleValue(List<Schema> parents, ConversionMode mode, JsonElement element,
			JsonElement defaultValue, String errorMessage) {

		if (mode == null)
			mode = ConversionMode.STRICT;

		return switch (mode) {
			case STRICT -> throw new SchemaConversionException(path(parents), element, mode, errorMessage);
			case LENIENT -> JsonNull.INSTANCE;
			case USE_DEFAULT -> defaultValue;
			case SKIP -> element;
		};
	}

	private ConvertorUtil() {
	}
}

File: kirun-js/src/engine/util/json/ConvertorUtil.ts
import { Schema } from '../../json/schema/Schema';
import { ConversionMode } from '../../json/schema/convertor/enums/ConversionMode';
import { SchemaConversionException } from '../../json/schema/convertor/exception/SchemaConversionException';
import { SchemaValidator } from '../../json/schema/validator/SchemaValidator';

export class ConvertorUtil {
    public static handleUnConvertibleValue(
        parents: Schema[],
        mode: ConversionMode | null,
        element: any,
        errorMessage: string,
    ): any | null {
        return this.handleUnConvertibleValueWithDefault(parents, mode, element, null, errorMessage);
    }

    public static handleUnConvertibleValueWithDefault(
        parents: Schema[],
        mode: ConversionMode | null,
        element: any,
        defaultValue: any,
        errorMessage: string,
    ): any | null {
        if (mode === null) {
            mode = ConversionMode.STRICT;
        }

        switch (mode) {
            case ConversionMode.STRICT:
                throw new SchemaConversionException(
                    SchemaValidator.path(parents),
                    element,
                    errorMessage,
                    mode,
                );
            case ConversionMode.LENIENT:
                return null;
            case ConversionMode.USE_DEFAULT:
                return defaultValue;
            case ConversionMode.SKIP:
                return element;
            default:
                throw new SchemaConversionException(
                    SchemaValidator.path(parents),
                    element,
                    'Invalid conversion mode',
                );
        }
    }

    private constructor() {}
}
